File: nodejs/base/worker/lib/jobs/process-job-assignment-helper.js
const { JobStatus, JobParameterBag, Exception } = require("@mcma/core");

class ProcessJobAssignmentHelper {
    constructor(dbTable, resourceManager, logger, workerRequest) {
        const jobAssignmentId = workerRequest.input.jobAssignmentId;

        let jobAssignment;
        let job;
        let profile;

        this.getTable = () => dbTable;
        this.getResourceManager = () => resourceManager;
        this.getLogger = () => logger;
        this.getRequest = () => workerRequest;

        this.getJobAssignmentId = () => jobAssignmentId;
        this.getJobAssignment = () => jobAssignment;
        this.getJob = () => job;
        this.getProfile = () => profile;
        this.getJobInput = () => job.jobInput;
        this.getJobOutput = () => job.jobOutput;

        this.initialize = async () => {
            jobAssignment = await this.updateJobAssignmentStatus(JobStatus.Running);

            job = await resourceManager.get(jobAssignment.job);

            profile = await resourceManager.get(job.jobProfile);

            job.jobOutput = jobAssignment.jobOutput || new JobParameterBag();
        };

        this.validateJob = () => {
            if (profile.inputParameters) {
                const missingInputParams = [];
                const jobInputKeys = job.jobInput && Object.keys(job.jobInput);
                for (const inputParam of profile.inputParameters) {
                    if (!jobInputKeys || !jobInputKeys.includes(inputParam.parameterName)) {
                        missingInputParams.push(inputParam.parameterName);
                    }
                }
                if (missingInputParams.length > 0) {
                    throw new Exception("One or more required input parameters are missing from the job: " + missingInputParams.join(", "));
                }
            }
        };

        this.complete = async (message) => {
            return await this.updateJobAssignment(
                ja => {
                    ja.status = JobStatus.Completed;
                    ja.statusMessage = message;
                    ja.jobOutput = job.jobOutput;
                },
                true);
        };

        this.fail = async (message) => {
            return await this.updateJobAssignment(
                ja => {
                    ja.status = JobStatus.Failed;
                    ja.statusMessage = message;
                    ja.jobOutput = job.jobOutput;
                },
                true);
        };

        this.cancel = async (message) => {
            return await this.updateJobAssignment(
                ja => {
                    ja.status = JobStatus.Canceled;
                    ja.statusMessage = message;
                    ja.jobOutput = job.jobOutput;
                },
                true);
        };

        this.updateJobAssignmentOutput = async () => {
            return await this.updateJobAssignment(ja => ja.jobOutput = job.jobOutput);
        };

        this.updateJobAssignmentStatus = async (status, statusMessage) => {
            return await this.updateJobAssignment(
                ja => {
                    ja.status = status;
                    ja.statusMessage = statusMessage;
                },
                true);
        };

        this.updateJobAssignment = async (update, sendNotification = false) => {
            if (typeof update !== "function") {
                throw new Exception("update must be a function that modifies the JobAssignment.");
            }
            jobAssignment = await dbTable.get(jobAssignmentId);
            if (!jobAssignment) {
                throw new Exception("JobAssignment with id '" + jobAssignmentId + "' not found.");
            }

            update(jobAssignment);

            jobAssignment.dateModified = new Date().toISOString();
            await dbTable.put(jobAssignmentId, jobAssignment);

            if (sendNotification) {
                await this.sendNotification();
            }

            return jobAssignment;
        };

        this.sendNotification = async () => {
            if (resourceManager) {
                await resourceManager.sendNotification(jobAssignment);
            }
        };
    }
}

module.exports = {
    ProcessJobAssignmentHelper
};

File: dotnet/base/Mcma.Worker/Jobs/ProcessJobAssignmentHelper.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Threading.Tasks;
using Mcma.Client;
using Mcma.Logging;
using Mcma.Data;

namespace Mcma.Worker
{
    public class ProcessJobAssignmentHelper<T> where T : Job
    {
        public ProcessJobAssignmentHelper(
            IDocumentDatabaseTable dbTable,
            IResourceManager resourceManager,
            WorkerRequestContext requestContext)
        {
            DbTable = dbTable;
            ResourceManager = resourceManager;
            RequestContext = requestContext;
            JobAssignmentId = requestContext.GetInputAs<ProcessJobAssignmentRequest>().JobAssignmentId;
        }

        public IDocumentDatabaseTable DbTable { get; }
        public IResourceManager ResourceManager { get; }
        public WorkerRequestContext RequestContext { get; }
        public string JobAssignmentId { get; }

        public ILogger Logger => RequestContext.Logger;
        public JobAssignment JobAssignment { get; private set; }
        public T Job { get; private set; }
        public JobProfile Profile { get; private set; }
        public JobParameterBag JobInput => Job.JobInput;
        public JobParameterBag JobOutput => Job.JobOutput;

        public async Task InitializeAsync()
        {
            JobAssignment = await UpdateJobAssignmentStatusAsync(JobStatus.Running);

            Job = await ResourceManager.ResolveResourceFromFullUrl<T>(JobAssignment.Job);
            Profile = await ResourceManager.ResolveResourceFromFullUrl<JobProfile>(Job.JobProfile);

            Job.JobOutput = new JobParameterBag();
        }

        public void ValidateJob(IEnumerable<string> supportedProfiles)
        {
            if (Profile.InputParameters == null) return;
            
            var missingInputParams =
                Profile.InputParameters.Where(p => !Job.JobInput.HasProperty(p.ParameterName)).Select(p => p.ParameterName).ToList();
            if (missingInputParams.Any())
                throw new Exception("One or more required input parameters are missing from the job: " + string.Join(", ", missingInputParams));
        }

        public Task<JobAssignment> CompleteAsync()
            => UpdateJobAssignmentAsync(
                ja =>
                {
                    ja.Status = JobStatus.Completed;
                    ja.JobOutput = Job?.JobOutput;
                },
                true);

        public Task<JobAssignment> FailAsync(ProblemDetail error)
            => UpdateJobAssignmentAsync(
                   ja =>
                   {
                       ja.Status = JobStatus.Failed;
                       ja.Error = error;
                       ja.JobOutput = Job?.JobOutput;
                   },
                   true);

        public Task<JobAssignment> CancelAsync()
            => UpdateJobAssignmentAsync(
                ja =>
                {
                    ja.Status = JobStatus.Canceled;
                    ja.JobOutput = Job?.JobOutput;
                },
                true);

        public Task<JobAssignment> UpdateJobAssignmentOutputAsync()
            => UpdateJobAssignmentAsync(ja => ja.JobOutput = JobOutput);

        public Task<JobAssignment> UpdateJobAssignmentStatusAsync(string status)
            => UpdateJobAssignmentAsync(
                ja => ja.Status = status,
                true);

        public async Task<JobAssignment> UpdateJobAssignmentAsync(Action<JobAssignment> update, bool sendNotification = false)
        {
            var jobAssignment = await GetJobAssignmentAsync();
            if (jobAssignment == null)
                throw new McmaException("JobAssignment with id '" + JobAssignmentId + "' not found");

            update(jobAssignment);

            jobAssignment.DateModified = DateTime.UtcNow;
            await DbTable.PutAsync(JobAssignmentId, jobAssignment);

            JobAssignment = jobAssignment;
            
            if (sendNotification)
                await SendNotificationAsync();

            return jobAssignment;
        }

        private async Task<JobAssignment> GetJobAssignmentAsync()
        {
            var jobAssignment = await DbTable.GetAsync<JobAssignment>(JobAssignmentId);

            foreach (var delay in new[] {5, 10, 15})
            {
                if (jobAssignment != null)
                    break;

                Logger?.Warn($"Failed to obtain job assignment from database table. Trying again in ${delay} seconds.");
                await Task.Delay(delay * 1000);
                jobAssignment = await DbTable.GetAsync<JobAssignment>(JobAssignmentId);
            }
            
            return jobAssignment;
        }

        public async Task SendNotificationAsync()
        {
            if (ResourceManager != null)
                await ResourceManager.SendNotificationAsync(JobAssignment, JobAssignment.NotificationEndpoint);
        }
    }
}
